atoma-confidential/src/sev_snp.rs

```diff
@@ -0,0 +1,91 @@
+use crate::ToBytes;
+
+use sev::{
+    attestation::{AttestationReport, AttestationReportSignature, CertTableEntry},
+    error::SevError,
+    firmware::Firmware,
+};
+
+use thiserror::Error;
+
+pub const SEV_SNP_REPORT_DATA_SIZE: usize = 64;
+
+type Result<T> = std::result::Result<T, SevError>;
+
+/// Generates an SEV-SNP attestation report for the given compute data.
+///
+/// This function takes a slice of bytes representing the attested data and returns
+/// an AttestationReport. The attested data must be exactly 64 bytes long,
+/// matching the SEV_SNP_REPORT_DATA_SIZE constant.
+///
+/// # Arguments
+///
+/// * `attested_data` - A slice of bytes containing the data to be attested (public key)
+///
+/// # Returns
+///
+/// * `Result<AttestationReport>` - On success, returns an AttestationReport.
+///                                 On failure, returns a SevError.
+///
+/// # Errors
+///
+/// Returns `SevError::InvalidAttestedDataSize` if the input data size is not 64 bytes.
+/// Returns `SevError::FailedVerification` if the attestation report verification fails.
+/// Returns `SevError::FailedToOpenFirmware` if the firmware interface cannot be opened.
+/// Returns `SevError::FailedToGetAttestationReport` if the attestation report cannot be generated.
+///
+/// # Example
+///
+/// ```
+/// let data_to_attest = [1u8; 64];
+/// match get_compute_data_attestation(&data_to_attest) {
+///     Ok(report) => println!("Attestation successful"),
+///     Err(e) => eprintln!("Attestation failed: {}", e),
+/// }
+/// ```
+pub fn get_compute_data_attestation(attested_data: &[u8]) -> Result<AttestationReport> {
+    if attested_data.len() != SEV_SNP_REPORT_DATA_SIZE {
+        return Err(SevError::InvalidAttestedDataSize(attested_data.len()));
+    }
+
+    let mut firmware = Firmware::open().map_err(SevError::FailedToOpenFirmware)?;
+
+    // Ask the PSP to generate an attestation report for the given data
+    let (attestation_report, cert_chain): (AttestationReport, Vec<CertTableEntry>) = firmware
+        .get_ext_report(Some(1), Some(attested_data), Some(1))
+        .map_err(SevError::FailedToGetAttestationReport)?;
+
+    // Verify the attestation report signature using the VCEK.
+    // This uses the crypto_nossl feature of the sev crate for detailed error handling and a pure Rust implementation.
+    (attestation_report, cert_chain)
+        .verify()
```

Comment:
I don't think we need to verify the attestation report signature at this point. This is better done at the client level

Reply:
When you say client, do you mean the atom-node on the same guest, or an atoma-node on a separate machine?
The reason why I included the verification is because the VCEK is only available on the same machine, and used for self-verfiying purposes (i.e. prove that the guest the AttestationReport was created from is a valid SEV-SNP guest since its signed with the VCEK, derived from the AMD Root key). I think the only way to verify that on a separate machine would be to also include the certificate chain, but I am not sure this is desireable.

https://docs.rs/sev/5.0.0/src/sev/certs/snp/mod.rs.html#40-49

Reply:
Yeah, I mean the atoma client (which can be running on my local machine) should verify attestations that are submitted on-chain, so that it knows if it can trust a given node or not.